llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -706,179 +626,266 @@ void EntryPointGroup::rebuildFromNames(const std::vector<std::string> &Names,
 
-    if (Aspects.empty())
-      return Ret + "-no-aspects";
+  // Accepts a callback, which should return a string based on provided
+  // function, which will be used as an entry points group identifier.
+  void registerRule(const std::function<std::string(Function *)> &Callback) {
+    Rules.emplace_back(Rule::RKind::K_Callback, Callback);
+  }
 
-    Ret += "-aspects";
-    for (int A : Aspects) {
-      Ret += "-" + std::to_string(A);
-    }
+  // Creates a simple rule, which adds a value of a string attribute into a
+  // resulting identifier.
+  void registerSimpleStringAttributeRule(StringRef AttrName) {
+    Rules.emplace_back(Rule::RKind::K_SimpleStringAttribute, AttrName);
+  }
 
-    if (UsesLargeGRF)
-      Ret += "-large-grf";
+  // Creates a simple rule, which adds one or another value to a resulting
+  // identifier based on a presence of a metadata on a function.
+  void registerSimpleFlagAttributeRule(StringRef AttrName,
+                                       StringRef IfPresentStr,
+                                       StringRef IfAbsentStr = "") {
+    Rules.emplace_back(Rule::RKind::K_FlagAttribute,
+                       std::tuple{AttrName, IfPresentStr, IfAbsentStr});
+  }
 
-    return Ret;
+  // Creates a simple rule, which adds one or another value to a resulting
+  // identifier based on a presence of a metadata on a function.
+  void registerSimpleFlagMetadataRule(StringRef MetadataName,
+                                      StringRef IfPresentStr,
+                                      StringRef IfAbsentStr = "") {
+    Rules.emplace_back(Rule::RKind::K_FlagMetadata,
+                       std::tuple{MetadataName, IfPresentStr, IfAbsentStr});
   }
 
-  static UsedOptionalFeatures getTombstone() {
-    UsedOptionalFeatures Ret;
-    Ret.IsTombstoneKey = true;
-    return Ret;
+  // Creates a rule, which adds a list of dash-separated integers converted
+  // into strings listed in a metadata to a resulting identifier.
+  void registerListOfIntegersInMetadataRule(StringRef MetadataName) {
+    Rules.emplace_back(Rule::RKind::K_IntegersListMetadata, MetadataName);
   }
 
-  static UsedOptionalFeatures getEmpty() {
-    UsedOptionalFeatures Ret;
-    Ret.IsEmpty = true;
-    return Ret;
+  // Creates a rule, which adds a list of sorted dash-separated integers
+  // converted into strings listed in a metadata to a resulting identifier.
+  void registerListOfIntegersInMetadataSortedRule(StringRef MetadataName) {
+    Rules.emplace_back(Rule::RKind::K_SortedIntegersListMetadata, MetadataName);
   }
 
 private:
-  // For DenseMap:
-  llvm::hash_code Hash = {};
-  bool IsTombstoneKey = false;
-  bool IsEmpty = false;
-
-public:
-  bool operator==(const UsedOptionalFeatures &Other) const {
-    // Tombstone does not compare equal to any other item
-    if (IsTombstoneKey || Other.IsTombstoneKey)
-      return false;
+  struct Rule {
+  private:
+    std::variant<StringRef, std::tuple<StringRef, StringRef, StringRef>,
```

Comment:
nit: could we make a type alias for what the tuple represents? tuple<stringref * 3> is not super intuitive to me

Reply:
Applied in 7ff7531215698fee717b3dadcb0987b3af1dee27. Not sure about the type alias name, please let me know if there are better suggestions

Reply:
Ok, I think I didn't get your original intent at first. `TupleOfThreeStringRef` is the same as `std::tuple<StringRef * 3>`, it does not represent domain-specific meaning of that data structure.

What about replacing the tuple with the following `struct` instead:
```
struct FlagRuleData {
  StringRef Name, IfPresentStr, IfAbsentStr;
};
```

This way it actually has some meaning and will allow to get rid of `std::get<N>`, which are not intuitive either. Instead we would access fields through meaningful names

Reply:
yeah that struct looks really good, thx

Reply:
Implemented that `struct` approach in e4b452e6142f5d1175758213a42f9821740dc85e